### ISSUE TITLE
Tricky to query the requires_grad() property of at::Tensor

### ISSUE THREAD

**Issue by meganset** [pytorch/pytorch#12383]
## 🐛 Bug

`auto t=at::tensor({1});`
 `t.requires_grad();`
_at::Error: requires_grad is not implemented for Tensor_

this is deliberately inhibited in core/TensorImpl.h 
not really a bug, but seems an innocent enough question for a tensor to ask of itself.

can go around via:
`t.options().requires_grad();`

but this doesn't get at the tensor's gradient setting:

`inline TensorOptions Tensor::options() const {`
 ` return TensorOptions().dtype(dtype())`
 `                       .device(device())`
 `                      .layout(layout())`
`                        .is_variable(is_variable());`
`}`

creating via torch instead means requires_grad() is permitted:
`auto v=torch::tensor({1.0}, at::requires_grad());`
`v.requires_grad();`
but querying via options is misleading:
`v.options().requires_grad();`
will be always false

## Additional context

(trying to sort out tensors from the at:: functions vs variables from torch:: in the new c++ api)

**Comment by goldsborough** [comment]
I think it's a bug that the `TensorOptions` returned by `Tensor::options` don't include `requires_grad`, thanks for reporting that. CC @ezyang 

To answer your more general question: In the C++ frontend, you never have to and never should use the `at::` namespace. All symbols you'd want to access in `at::` are exported to `torch::`. By ignoring the `at::` namespace entirely you won't have to deal with the distinction between tensors and variables (which is the root problem here) at all. Just see `torch::Tensor` as the equivalent to `torch.Tensor` in Python -- it's always a variable.

**Comment by ezyang** [comment]
> I think it's a bug that the TensorOptions returned by Tensor::options don't include requires_grad, thanks for reporting that.

Actually, I'm not sure... say you write something like this:

```
void f(Tensor x) {
  auto y = at::zeros({2,2}, x.options());
  do_stuff_to_it_mutably(y);
  return x + y;
}
```

You almost definitely didn't want to make `y` a leaf variable that accumulates grad. So `requires_grad` feels a bit different than scalar types and device, at least with the idiom here.

But yes, I agree that everything would be better if Tensor and Variable merge in C++, and we're working on making this happen. You will be a much happier panda if you just use variables exclusively.

**Comment by meganset** [comment]
Closing this as the evolution of TensorOptions and Tensor clears up most of the trouble
